src/shmem_lock.h

```diff
@@ -93,6 +94,8 @@ shmem_internal_set_lock(long *lockp)
 
             SHMEM_WAIT(&(lock->data), lock_cur.data);
         }
+    } else {
+        shmem_internal_membar_load();
```

Comment:
I will surely forget the reason for this memory barrier.  Could you add a comment to the effect: Lock was acquired immediately without calling SHMEM_WAIT, which provides memory ordering.  Therefore, issue a load fence to ensure memory ordering.